here-naksha-app-service/src/main/java/com/here/naksha/app/service/http/tasks/WriteFeatureApiTask.java

```diff
@@ -228,6 +238,202 @@ protected void init() {}
     }
   }
 
+  // TODO this might change if naksha_plpgsql.sql gets updated and the message is modified
+  private static final Pattern FEATURE_ID_FROM_ERR = Pattern.compile("The feature \\s*'([^']*)' uuid");
+  private static final int MAX_RETRY_ATTEMPT = 5;
+
+  private @NotNull XyzResponse executePatchFeatureById() throws JsonProcessingException {
+
+    final XyzFeature featureFromRequest = singleFeatureFromRequestBody();
+
+    final String spaceId = ApiParams.extractMandatoryPathParam(routingContext, SPACE_ID);
+    final String featureId = ApiParams.extractMandatoryPathParam(routingContext, FEATURE_ID);
+    final QueryParameterList queryParams = queryParamsFromRequest(routingContext);
+    final List<String> addTags = extractParamAsStringList(queryParams, ADD_TAGS);
+    final List<String> removeTags = extractParamAsStringList(queryParams, REMOVE_TAGS);
+
+    // Validate parameters
+    if (!featureId.equals(featureFromRequest.getId())) {
+      return verticle.sendErrorResponse(
+          routingContext,
+          XyzError.ILLEGAL_ARGUMENT,
+          "URI path parameter featureId is not the same as id in feature request body.");
+    }
+
+    // as applicable, modify features based on parameters supplied
+    addTagsToFeature(featureFromRequest, addTags);
+    removeTagsFromFeature(featureFromRequest, removeTags);
+
+    final List<XyzFeature> featuresFromRequest = new ArrayList<>();
+    featuresFromRequest.add(featureFromRequest);
+    List<String> featureIds = new ArrayList<>();
+    featureIds.add(featureId);
+    return attemptFeaturesPatching(spaceId, featureIds, featuresFromRequest, HttpResponseType.FEATURE, 0);
+  }
+
+  private XyzResponse attemptFeaturesPatching(
```

Comment:
I understand that this operation is complex but this method is too long. Moreover, there are significant actions here that can be extracted to separate private methods - it will reduce complexity for reader and maybe you'll be able to find common actions that could be later on reused in different places.

There are many approaches and tricks to deal with such complex methods - you can take a look at this small blog entry that names few of them: https://refactoring.guru/smells/long-method
A bit longer, but definitely worth reading piece is Martin Fowler's Refactoring book (IMO it's a must-read): https://martinfowler.com/books/refactoring.html

I'll be happy to asssist you in that process if you'd like to :)

Reply:
Couple of other resources:
- [this thread with great discussion](https://softwareengineering.stackexchange.com/questions/133404/what-is-the-ideal-length-of-a-method-for-you)
- [SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle)
- [yet another awesome book that discusses these topcis](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) - Martin's Clean Code